test/trento/domain/sap_system/sap_system_test.exs

```diff
@@ -4202,6 +4210,365 @@ defmodule Trento.SapSystemTest do
+      host_id = Faker.UUID.v4()
+      absent_app_instance_number = "02"
+      present_app_instance_number = "03"
+      absent_app_absent_at = DateTime.utc_now()
+
+      initial_events = [
+        build(
+          :database_registered_event,
+          sap_system_id: sap_system_id,
+          sid: sid
+        ),
+        build(
+          :database_instance_registered_event,
+          sap_system_id: sap_system_id,
+          sid: sid,
+          host_id: host_id,
+          instance_number: "01",
+          system_replication: nil,
+          system_replication_status: nil
+        ),
+        build(:application_instance_registered_event,
+          sap_system_id: sap_system_id,
+          sid: sid,
+          host_id: host_id,
+          instance_number: absent_app_instance_number,
+          features: "MESSAGESERVER"
+        ),
+        build(
+          :application_instance_marked_absent_event,
+          sap_system_id: sap_system_id,
+          host_id: host_id,
+          instance_number: absent_app_instance_number,
+          absent_at: absent_app_absent_at
+        ),
+        build(:application_instance_registered_event,
+          sap_system_id: sap_system_id,
+          sid: sid,
+          host_id: host_id,
+          instance_number: present_app_instance_number,
+          features: "ABAP"
+        )
+      ]
+
+      absent_at = DateTime.utc_now()
+
+      assert_events_and_state(
+        initial_events,
+        [
+          %MarkApplicationInstanceAbsent{
+            instance_number: absent_app_instance_number,
+            host_id: host_id,
+            sap_system_id: sap_system_id,
+            absent_at: absent_at
+          },
+          %MarkApplicationInstanceAbsent{
+            instance_number: present_app_instance_number,
+            host_id: host_id,
+            sap_system_id: sap_system_id,
+            absent_at: absent_at
+          }
+        ],
+        [
+          %ApplicationInstanceMarkedAbsent{
+            instance_number: present_app_instance_number,
+            host_id: host_id,
+            sap_system_id: sap_system_id,
+            absent_at: absent_at
+          }
+        ],
+        fn state ->
+          assert %SapSystem{
+                   application: %SapSystem.Application{
+                     sid: ^sid,
+                     instances: [
+                       %SapSystem.Instance{
+                         instance_number: ^present_app_instance_number,
+                         absent_at: ^absent_at
+                       },
+                       %SapSystem.Instance{
+                         instance_number: ^absent_app_instance_number,
+                         absent_at: ^absent_app_absent_at
+                       }
+                     ]
+                   }
+                 } = state
+        end
+      )
+    end
+
+    test "receiving a 'register instance' command for an already-registered database instance sets instance as present if absent" do
```

Comment:
I would rephrase this to:
```
should mark as present and already registered absent database instance
```
We use the `should` wording at the beginning of each test by default